server/src/lib/server.rs

```diff
@@ -649,51 +652,131 @@ async fn instance_state_put(
+            },
             None => (
                 serial.read_source(&mut output, actx).fuse(),
                 Fuse::terminated(),
             ),
             Some(r) => (
                 Fuse::terminated(),
-                ws_sink.send(Message::binary(&output[r.clone()])).fuse(),
+                futures::stream::iter(
+                    ws_sinks
+                        .iter_mut()
+                        .zip(std::iter::repeat(Vec::from(&output[r.clone()]))),
+                )
+                .for_each_concurrent(4, |(ws, bin)| {
+                    ws.send(Message::binary(bin)).map(|_| ())
+                })
+                .right_future()
+                .fuse(),
             ),
         };
+
         let (ws_recv, uart_write) = match &cur_input {
-            None => (ws_stream.next().fuse(), Fuse::terminated()),
+            None => (
+                futures::stream::iter(ws_streams.iter_mut().enumerate())
+                    .for_each_concurrent(4, |(i, ws)| {
+                        // if we don't `move` below, rustc says that `i` (which is usize: Copy (!))
+                        // is borrowed. but if we move without making this explicit reference here,
+                        // it moves send_ch into the closure.
+                        let ch = &send_ch;
+                        ws.next().then(move |msg| ch.send((i, msg))).map(|_| ())
+                    })
+                    .fuse(),
+                Fuse::terminated(),
+            ),
             Some((data, consumed)) => (
                 Fuse::terminated(),
                 serial.write_sink(&data[*consumed..], actx).fuse(),
             ),
         };
 
+        let recv_ch_fut = recv_ch.recv();
+
+        let detach = detach_opt.as_mut().unwrap_or(&mut done_detaching_recv);
+
         tokio::select! {
             // Poll in the order written
             biased;
 
             // It's important we always poll the detach channel first
             // so that a constant stream of incoming/outgoing messages
             // don't cause us to ignore a detach
-            _ = &mut detach => {
-                info!(log, "Detaching from serial console");
-                let close = CloseFrame {
-                    code: CloseCode::Policy,
-                    reason: Cow::Borrowed("serial console was detached"),
-                };
-                ws_sink.send(Message::Close(Some(close))).await?;
-                break;
+            _ = detach => {
+                if !detaching {
+                    detaching = true;
+                    detach_opt = None; // can't re-poll a consumed oneshot channel
+                    info!(log, "Detaching from serial console, disconnecting all clients");
+                } else {
+                    info!(log, "Finished detaching all serial clients");
+                    break;
+                }
+            }
+
+            new_ws = websocks_recv.recv() => {
+                if !detaching {
```

Comment:
Worth adding a more graceful close here for an client attempting to connect while detaching?